pkg/generator/specification.go

```diff
@@ -126,43 +126,44 @@ func (s *SpecificationGenerator) generateIATPerGranularity(numberOfInvocations i
 }
 
 // GenerateIAT generates IAT according to the given distribution. Number of minutes is the length of invocationsPerMinute array
-func (s *SpecificationGenerator) generateIAT(invocationsPerMinute []int, iatDistribution common.IatDistribution, shiftIAT bool, granularity common.TraceGranularity) (common.IATMatrix, common.ProbabilisticDuration) {
-	var IAT [][]float64
+func (s *SpecificationGenerator) generateIAT(invocationsPerMinute []int, iatDistribution common.IatDistribution,
+	shiftIAT bool, granularity common.TraceGranularity) (common.IATArray, []int, common.ProbabilisticDuration) {
+
+	var IAT []float64
+	var perMinuteCount []int
 	var nonScaledDuration []float64
 
 	numberOfMinutes := len(invocationsPerMinute)
 	for i := 0; i < numberOfMinutes; i++ {
 		minuteIAT, duration := s.generateIATPerGranularity(invocationsPerMinute[i], iatDistribution, shiftIAT, granularity)
 
-		IAT = append(IAT, minuteIAT)
+		IAT = append(IAT, minuteIAT...)
+		perMinuteCount = append(perMinuteCount, len(minuteIAT))
```

Comment:
Why do you have separate array here? Shouldn't it be just the array that we already got from the trace?

Reply:
It ended up not being the same array. I noticed a wired behavior of the specification generator since introduction of these changes. For example, if we say we want to create 5 invocations within a minute, we get an array of length 6. The elements are 0, 12_000_000, 12_000_000, 12_000_000, 12_000_000, 12_000_000. I think this is caused by line 122.

![image](https://github.com/vhive-serverless/invitro/assets/16463659/7ff7bc13-f43d-44dc-83b7-69fd02b50903)

Reply:
I've made this changes to have the same format of IATs for with and without shifted invocations to have the driver logic the same independent of IAT that we've chosen. Everytime, except empty minutes, it creates one more number than there are invocations. First one being the interval between beginning of the minute and first invocation, all other invocations happen after delay stated in next elements of the array. Last one would just align the sum to 60s without invocation after.

Since we are flattening the array, maybe we should merge the last element of previous minute with first of the next one. After that, we won't need this additional element for each minute. But that would change the logic of skipping the minute in driver. And there wouldn't be possible the sleeps to the end of the minute in the driver, minute counting should be careful.